include/bout/operators_di.hxx

```diff
@@ -17,7 +17,7 @@
 /// @param i  The point where the result is calculated
 ///
 inline BoutReal DDX_C2(const Field3D &f, const DataIterator &i) {
-  return (f[i.xp()] - f[i.xm()])/(2.*mesh->coordinates()->dx[i]);
+  return (f[i.xp()] - f[i.xm()])/(2.*mesh->coordinates(f.getLocation())->dx[i]);
```

Comment:
I know this might be giving a bit too much authority to Field?s but could we have a `Field?D.getCoordinates()` method that would return `fieldmesh->coordinates(location)`? This might make statements like this clearer as we'd replace `mesh->coordinates(f.getlocation())->dx[i]` with `f.getCoordinates()->dx[i]`. A further "bonus" is that it would mean we'd be using the field's mesh everywhere like this rather than the global one (i.e. strictly `mesh->coordinates(f.getlocation())->dx[i]`  should really be `f.getMesh()->coordinates(f.getlocation())->dx[i]`

Reply:
Yes, agree this would be good. It is adding a bit to Field, but would make the code more readable.

Reply:
Also in this case some way is needed to make this lookup faster; in this operator the coordinates are fetched for every point in the inner loop.

Reply:
If the Field*D stored a coordinates pointer initialized set to `fieldmesh->coordinates(location)` instead of calling through fieldmesh, `f.getCoordinates()` could avoid the lookup. I guess it could then be inlined by the compiler as well?

Reply:
Good idea; it certainly should be a lot faster. Would just need to ensure that the coordinates pointer was updated when location was changed.

Reply:
I'm happy to have a stab at implementing this unless anybody else has started?

Reply:
Slightly more complicated than first thought -- has knock-on effects for tests etc. as now all staggering tests require a valid mesh/coordinates to exist which seems to be causing problems. Easy to get into an infinite loop with obvious design (creating coordinates triggers fields created which then tries to create coordinates etc.) but should be possible to address with a reasonable interface.

Reply:
Design question : Should `getCoordinates()` will return the coordinates pointer at the correct location, should we also provide `getCoordinates(CELL_LOC loc)` that just returns `getMesh()->coordinates(loc)` as a convenience function? It would tidy some bits of code but might not be what we want.

Reply:
#1266 should deal with most of these changes.